### PR TITLE
Support TraceLoggingBinary for map find, update and delete.

### PR DESCRIPTION
**Pull request by shpalani**
## Description
This PR contains 
1. EBPF_LOG_MESSAGE_BINARY tracing log with TraceLoggingBinary to print binary data.
    #define _EBPF_LOG_MESSAGE_BINARY(trace_level, keyword, message, data, data_size)
2. Tracing added to the following functions:
 - ebpf_map_find_entry
 - ebpf_map_update_entry
 - ebpf_map_delete_entry
 
 3.  Reduced NET_EBPF_EXT_STACK_EXPANSION_SIZE to 4kb for all images.
4.  Enabled whole program optimization for tracelog.c
5. Added stack expansion for net_ebpf_extension_sock_addr_authorize_recv_accept_classify
  
This tracing aims to instrument the map invocations and tail calls in the drivers.

## Testing
_Do any existing tests cover this change? N/A
Are new tests needed? No, as this is a trace log.

Logs:
[0]0004.15E8::2024/09/17-18:08:02.977888400 [EbpfForWindowsProvider]{"Message":"Tail call program","string":"xxxx","meta":{"provider":"EbpfForWindowsProvider","event":"EbpfGenericMessage","time":"2024-09-18T01:08:02.9778884Z","cpu":0,"pid":4,"tid":5608,"channel":11,"level":5,"keywords":"0x80"}}
[0]0004.15E8::2024/09/17-18:08:02.977892100 [EbpfForWindowsProvider]{"Message":"Tail call program","string":"xxxx","meta":{"provider":"EbpfForWindowsProvider","event":"EbpfGenericMessage","time":"2024-09-18T01:08:02.9778921Z","cpu":0,"pid":4,"tid":5608,"channel":11,"level":5,"keywords":"0x80"}}
[0]0004.15E8::2024/09/17-18:08:02.977894000 [EbpfForWindowsProvider]{"Message":"**Map update**","string":"xxxx","meta":{"provider":"EbpfForWindowsProvider","event":"EbpfGenericMessage","time":"2024-09-18T01:08:02.9778940Z","cpu":0,"pid":4,"tid":5608,"channel":11,"level":5,"keywords":"0x40"}}
[0]0004.15E8::2024/09/17-18:08:02.977894700 [EbpfForWindowsProvider]{"Message":"Key","(data).Length":12,"(data)":"0x0A0101030403020100010600","meta":

Note: Program name and map name are masked with xxxx in the logs.

## Documentation

_Is there any documentation impact for this change? No

## Installation

_Is there any installer impact for this change? No

### CODE REVIEW

File: libs/execution_context/ebpf_program.c [hunk 1]
```diff
@@ -1559,6 +1559,13 @@ ebpf_program_invoke(
             ebpf_program_entry_point_t function_pointer;
             function_pointer = (ebpf_program_entry_point_t)(current_program->code_or_vm.code.code_pointer);
             *result = (function_pointer)(context);
+#ifdef _DEBUG
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:1562]: There are currently no ifdef _DEBUG lines anywhere in this directory.  Other log calls check the state inside the macro, why are ifdefs needed here?

File: libs/shared/ebpf_tracelog.h [hunk 2]
```diff
@@ -303,6 +303,20 @@ extern "C"
...
     }
 
+    void
+    ebpf_log_message_binary(
+        ebpf_tracelog_level_t trace_level,
+        ebpf_tracelog_keyword_t keyword,
+        _In_z_ const char* message,
+        _In_z_ const char* type,
+        _In_ const cxplat_utf8_string_t* name,
+        const void* data,
```
Comment (dthaler) [inline libs/shared/ebpf_tracelog.h:313]: missing SAL annotation on this line

File: libs/shared/tracelog.c [hunk 3]
```diff
@@ -976,6 +976,89 @@ __declspec(noinline) void ebpf_log_message_uint64_uint64(
...
+        ebpf_assert(!"Invalid keyword");                                                                          \
+        break;                                                                                                    \
+    }
+__declspec(noinline) void ebpf_log_message_binary(
+    ebpf_tracelog_level_t trace_level,
+    ebpf_tracelog_keyword_t keyword,
+    _In_z_ const char* message,
+    _In_z_ const char* type,
+    _In_ const cxplat_utf8_string_t* name,
+    const void* data,
```
Comment (dthaler) [inline libs/shared/tracelog.c:1034]: Missing SAL annotation on this line

File: libs/execution_context/ebpf_program.c (hunk 1 above)
Comment (shpalani) [reply libs/execution_context/ebpf_program.c:1562]: Thanks. The intent was to add the trace only in DEBUG image with INFO level. Did you mean to change the log level to LEVEL_VERBOSE, and remove #ifdef?

Comment (dthaler) [reply libs/execution_context/ebpf_program.c:1562]: yes that would be better

**Review by dthaler — DISMISSED** [review]

File: libs/execution_context/ebpf_maps.c [hunk 4]
```diff
@@ -2671,6 +2685,12 @@ ebpf_map_delete_entry(_In_ ebpf_map_t* map, size_t key_size, _In_reads_(key_size
         return EBPF_OPERATION_NOT_SUPPORTED;
     }
 
+    if (flags & EBPF_MAP_FLAG_HELPER && map->ebpf_map_definition.key_size != 0) {
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:2688]:
```suggestion
    if ((flags & EBPF_MAP_FLAG_HELPER) && map->ebpf_map_definition.key_size != 0) {
```
for consistency with line 2619

**Review by dthaler — APPROVED** [review]
Approving but see nit

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Review by dthaler — DISMISSED** [review]

**Review by shankarseal — DISMISSED** [review]

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 5]
```diff
@@ -1570,7 +1570,8 @@ net_ebpf_extension_sock_addr_authorize_recv_accept_classify(
         goto Exit;
     }
 
-    program_result = net_ebpf_extension_hook_invoke_programs(sock_addr_ctx, &filter_context->base, &result);
+    program_result =
+        net_ebpf_extension_hook_expand_stack_and_invoke_programs(sock_addr_ctx, &filter_context->base, &result);
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:1574]: I would suggest that `net_ebpf_extension_hook_invoke_programs` takes a new `flags` bitmask parameter. One of the bits can control if it should internally call expand stack. Only sock_addr should pass TRUE in that call. No need in other providers.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 6]
```diff
@@ -136,13 +140,19 @@ if ($VerbosePreference -eq 'Continue') {
 }
 
 # Get the available free space before test start (useful in investigating dump file creation failures)
-$BeforeTestFreeSpaceGB = GetDriveFreeSpaceGB -DriveSpecification $Env:SystemDrive
+try {
+    $BeforeTestFreeSpaceGB = GetDriveFreeSpaceGB -DriveSpecification $Env:SystemDrive
+} catch {
+    Write-Log "Error getting available disk space: $_"
+    $BeforeTestFreeSpaceGB = "Unknown"
+    // Continue with the test.
```
Comment (shankarseal) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:148]: Typo? PS comments are `#`.

File: scripts/config_test_vm.psm1 [hunk 7]
```diff
@@ -324,11 +324,25 @@ function ArchiveKernelModeDumpOnVM
...
+                    Compress-Archive `
+                        -Path "$KernelModeDumpFileSourcePath\*.dmp" `
+                        -DestinationPath "$KernelModeDumpFileDestinationPath\km_dumps.zip" `
+                        -CompressionLevel Fastest `
+                        -Force
+                    break
+                } catch {
+                    $ErrorMessage = "*** ERROR *** Failed to compress kernel mode dump files: $_. Retry"
+                    Write-Output $ErrorMessage
+                    Start-Sleep -seconds 5
```
Comment (shankarseal) [inline scripts/config_test_vm.psm1:342]: Suggest: `Start-Sleep -seconds 5 * $retry_count` and start `$retry_count` at 1.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 5 above)
Comment (shpalani) [reply netebpfext/net_ebpf_ext_sock_addr.c:1574]: Available stack expansion api KeExpandKernelStackAndCalloutEx needs a callout. Hence taking the existing approach, which uses KeExpandKernelStackAndCalloutEx with a callout to invoke net_ebpf_extension_hook_invoke_programs.

File: netebpfext/net_ebpf_ext_hook_provider.c [hunk 8]
```diff
@@ -4,7 +4,7 @@
 #include "ebpf_extension_uuids.h"
 #include "net_ebpf_ext_hook_provider.h"
 
-#define NET_EBPF_EXT_STACK_EXPANSION_SIZE 1024 * 10
+#define NET_EBPF_EXT_STACK_EXPANSION_SIZE 1024 * 16
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_hook_provider.c:7]: I would still prefer 4K for release and 16K for debug.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 5 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_sock_addr.c:1574]: OK - after we discussed this more, please add these to bind and sock_ops as well.

File: netebpfext/net_ebpf_ext_hook_provider.c [hunk 9]
```diff
@@ -4,7 +4,11 @@
 #include "ebpf_extension_uuids.h"
 #include "net_ebpf_ext_hook_provider.h"
 
-#define NET_EBPF_EXT_STACK_EXPANSION_SIZE 1024 * 10
+#ifdef _DEBUG
+#define NET_EBPF_EXT_STACK_EXPANSION_SIZE 1024 * 20
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_hook_provider.c:8]: The latest crash dump shows the stack is about 23K bytes. With the frames above the call to the ETW API around 9K. I suggest giviing it one more try with increasing the stack by 32K for debug.

File: netebpfext/net_ebpf_ext_hook_provider.c (hunk 8 above)
Comment (shpalani) [reply netebpfext/net_ebpf_ext_hook_provider.c]: In latest, it is 4K.

**Review by Alan-Jowett — APPROVED** [review]

**Review by mikeagun — APPROVED** [review]

**Pull request merged**